homeassistant/components/camera/dlink.py

```diff
@@ -0,0 +1,341 @@
+from homeassistant.components.camera import Camera
+from homeassistant.loader import get_component
+# pylint: disable=import-error
+from bs4 import BeautifulSoup
+
+_LOGGER = logging.getLogger(__name__)
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=25)
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """ Find and return Vera lights. """
+    if not validate_config(
+            {DOMAIN: config},
+            {DOMAIN: ['base_url', CONF_USERNAME, CONF_PASSWORD]},
+            _LOGGER):
+        return None
+
+    model_map = {
+        'dcs-930l': DlinkCameraDcs930l
+    }
+
+    family_map = {
+        'dcs': DlinkCameraDcs930l
+    }
+
+    # To load up the camera component we first check if there is component
+    # for the specific model if not we then check the specified family,
+    # if that also fails we just load the generic comonent for this brand.
+
+    camera_class = DlinkCamera
+    model = config.get('model', False)
+    if model:
+        model = model.lower()
+        camera_class = model_map.get(model, DlinkCamera)
+
+    family = config.get('family', False)
+    if family and camera_class == DlinkCamera:
+        family = family.lower()
+        camera_class = family_map.get(family, DlinkCamera)
+
+    camera = camera_class(hass, config)
+    cameras = [camera]
+
+    add_devices_callback(cameras)
+
+
+# pylint: disable=too-many-public-methods
+class DlinkCamera(Camera):
+    """ A generic D-Link camera class """
+    def __init__(self, hass, device_info):
+        """ Initialise the generic D-Link camera class. """
+        # pylint: disable=missing-super-argument
+        super().__init__(hass, device_info)
+        self._still_image_url = device_info.get('still_image_url', 'image.jpg')
+
+
+# pylint: disable=too-many-public-methods
+# pylint: disable=too-many-instance-attributes
+class DlinkCameraDcs930l(Camera):
+    """ A class designed for the D-Link DCS-930L, it may also work with other
+    related devices """
+    def __init__(self, hass, device_info):
+        """ Initialise the D-Link DCS-930L camera component. """
+        # pylint: disable=missing-super-argument
+        super().__init__(hass, device_info)
+        self._is_motion_detection_supported = True
+        self._is_ftp_upload_supported = True
+        self._can_auto_configure_ftp = True
+        self._still_image_url = device_info.get('still_image_url', 'image.jpg')
+
+        # Holds the form data so we can post updates back to the web UI
+        self._web_ui_form_data = {}
+
+    def refesh_all_settings_from_device(self):
+        """ Overrides the base class method that retrieved all setting
+            from the device. """
+        self.get_all_settings()
+
+    def get_all_settings(self):
+        """ Pull all the settings from the camera, there is no API so it's
+        dirty screen scraping time. """
+
+        res = requests.get(
```

Comment:
All the screen scraping and interacting with DLink camera logic should be extracted into a separate Python library just like vera / nzbclient.